File: go-sort/radix_sort.go
package gosort

func RadixSort(arr []int) {
	maxValue := GetArrayMaxValue(arr)
	for r := 1; maxValue/r > 0; r *= 10 {
		radixSort(arr, r)
	}
}

func radixSort(arr []int, r int) {
	temp := make([]int, len(arr))
	buckets := make([]int, 10)
	for _, v := range arr {
		buckets[(v/r)%10]++
	}
	for i := 1; i < len(buckets); i++ {
		buckets[i] += buckets[i-1]
	}
	for i := len(arr) - 1; i >= 0; i-- {
		temp[buckets[(arr[i]/r)%10]-1] = arr[i]
		buckets[(arr[i]/r)%10]--
	}
	copy(arr, temp)
}

File: ts-sort/radixSort.ts
import { getMaxValueFromArray } from './common'

export default function radixSort(arr: number[]) {
  const maxValue = getMaxValueFromArray(arr)
  for (let r = 1; Math.trunc(maxValue / r) > 0; r *= 10) {
    _radixSort(arr, r)
  }
}

function _radixSort(arr: number[], r: number) {
  const temp = Array.from<number>({ length: arr.length }).fill(0)
  const buckets = Array.from<number>({ length: 10 }).fill(0)
  for (const v of arr) {
    const index = Math.trunc(v / r) % 10
    buckets[index]!++
  }
  for (let i = 1; i < buckets.length; i++) {
    buckets[i]! += buckets[i - 1]!
  }
  for (let i = arr.length - 1; i >= 0; i--) {
    const index = Math.trunc(arr[i]! / r) % 10
    temp[buckets[index]! - 1] = arr[i]!
    buckets[index]!--
  }
  for (let i = 0; i < arr.length; i++) {
    arr[i] = temp[i]!
  }
}
